provisioning/miniprov/hare_mp/types.py

```diff
@@ -87,22 +92,22 @@ class M0Clients(DhallTuple):
 
 @dataclass(repr=False)
 class DisksDesc(DhallTuple):
-    meta_data: Maybe  # [str]
-    data: DList
+    meta_data: Maybe[Text]
+    data: DList[Text]
 
 
 @dataclass(repr=False)
 class M0ServerDesc(DhallTuple):
-    runs_confd: Maybe  # [bool]
+    runs_confd: Maybe[bool]
     io_disks: DisksDesc
 
 
 @dataclass(repr=False)
 class NodeDesc(DhallTuple):
     hostname: Text
     data_iface: Text
-    data_iface_type: Maybe  # [Protocol]
-    io_disks: DList  # [str]
+    data_iface_type: Maybe[Protocol]
+    io_disks: DList[Text]
```

Comment:
Is `Text` really needed instead of `str`?  :)
From documentation i understood that it is just for compatibility with Python 2.

Reply:
Yes, it is needed.

The fact is that Text, NodeDesc, Protocol and other guys are a kind of AST objects. Their main aim is to stringify themself into Dhall expressions. Use case: generate CDF file according to some values in ConfStore.  In order to generate a valid (and type safe) CDF file, we reuse Dhall types. In order to generate Dhall expressions we use objects like this.

So the main problem is that Dhall strings are enclosed into double quotes (`"like this"`). While str() for any string in Python will yield a string within single quotes. Even in Python 3

```

Python 3.6.8 (default, Aug  7 2019, 17:28:10)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> t = 'Some text goes here'
>>> str(t)
'Some text goes here'
>>>
```

Reply:
So, I we create our own class whose `__str__` implementation prints double-quotes, right? 
I still don't understand how usage of  `Text` type hint helps in this situation :(

Reply:
Exactly. Quote from the tests:

```python
    def test_m0server_with_disks(self):
        val = M0ServerDesc(
            runs_confd=Maybe(True, 'Bool'),
            io_disks=DisksDesc(
                meta_data=Maybe(None, 'Text'),
                data=DList([Text('/disk1'), Text('/disk2')], 'test')))
        self.assertEqual(
            '{ runs_confd = Some (True), io_disks = { meta_data = None Text, data = ["/disk1", "/disk2"] } }',
            str(val))

```

Note how `data = [ ... ]` array looks like in the assert.

Reply:
Ah, now i got it. I mixed it with https://docs.python.org/3/library/typing.html#typing.Text
Didn't notice separate Text implementation :)